Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE CPP #-}
+
+module Math.NumberTheory.Primes.Factorisation.QuadraticSieve
+  ( quadraticSieve
+  , findPairs
+  ) where
+
+#if __GLASGOW_HASKELL__ < 803
+import Data.Semigroup
+#endif
+import qualified Data.List as L
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as MV
+import qualified Data.IntMap as I
+import qualified Data.IntSet as S
+import qualified Math.NumberTheory.Primes.IntSet as PS
+import Control.Monad
+import Control.Monad.ST
+import Data.Maybe
+import Data.Bifunctor
+import Math.NumberTheory.Roots
+import Math.NumberTheory.Primes
+import Math.NumberTheory.Moduli.Sqrt
+import Math.NumberTheory.Utils.FromIntegral
+
+data SignedPrimeIntSet = SignedPrimeIntSet
+  { sign :: !Bool
+  , primeSet :: !PS.PrimeIntSet
+  } deriving (Show)
+
+data BoolOrPrime = Bool !Bool | PrimeInt !(Prime Int)
+
+insert :: Prime Int -> SignedPrimeIntSet -> SignedPrimeIntSet
+insert prime (SignedPrimeIntSet s ps) = SignedPrimeIntSet s (prime `PS.insert` ps)
+
+nonZero :: SignedPrimeIntSet -> Maybe BoolOrPrime
+nonZero (SignedPrimeIntSet s ps) = case PS.minView ps of
+  Just (prime, _)  -> Just (PrimeInt prime)
+  Nothing          -> if s then Just (Bool True) else Nothing
+
+member :: BoolOrPrime -> SignedPrimeIntSet -> Bool
+member value (SignedPrimeIntSet s ps) = case value of
+  Bool b     -> s == b
+  PrimeInt p -> p `PS.member` ps
+
+xor :: SignedPrimeIntSet -> SignedPrimeIntSet -> SignedPrimeIntSet
+xor (SignedPrimeIntSet s1 ps1) (SignedPrimeIntSet s2 ps2) = SignedPrimeIntSet (s1 /= s2) ((ps1 PS.\\ PS.unPrimeIntSet ps2) <> (ps2 PS.\\ PS.unPrimeIntSet ps1))
+
+-- Given an odd positive composite Integer n and Int parameters b and t,
+-- the Quadratic Sieve attempt to decompose n into smaller factors p and q.
+quadraticSieve :: Integer -> Int -> Int -> Integer
+quadraticSieve n b t = findFactor n $ findPairs n b t
+
+findFactor :: Integer -> [(Integer, Integer)] -> Integer
+findFactor n pairs = case L.find (\(x, y) -> gcd (x - y) n /= 1 && gcd (x - y) n /= n) pairs of
+  Just (x, y) -> gcd (x - y) n
+  Nothing     -> error "Parameters are not large enough."
+
+findPairs :: Integer -> Int -> Int -> [(Integer, Integer)]
```

Comment:
This function is exported, but does not have a Haddock documentation.

Reply:
Thank you for taking time to review the code. What do you mean by this? Do I need to create a separate file?

Reply:
Well there are several possibilities, including using a preprocessor, but I think something like this would do:
https://www.reddit.com/r/haskell/comments/8w90ia/test_non_public_functions/e1tocry?utm_source=share&utm_medium=web2x

Reply:
Eventually all these functions will be hidden behind `Math.NumberTheory.Primes.factorise`, but for the time being it is more convenient to have them exported.